internal/fs/sd_windows.go

```diff
@@ -0,0 +1,469 @@
+		s += getPrivilegeName(p)
+		s += `"`
+	}
+	if backupPrivilegeError != nil {
+		s += " backupPrivilegeError:" + backupPrivilegeError.Error()
+	}
+	if restorePrivilegeError != nil {
+		s += " restorePrivilegeError:" + restorePrivilegeError.Error()
+	}
+	return s
+}
+
+func mapPrivileges(names []string) ([]uint64, error) {
+	privileges := make([]uint64, 0, len(names))
+	privNameMutex.Lock()
+	defer privNameMutex.Unlock()
+	for _, name := range names {
+		p, ok := privNames[name]
+		if !ok {
+			err := lookupPrivilegeValue("", name, &p)
+			if err != nil {
+				return nil, err
+			}
+			privNames[name] = p
+		}
+		privileges = append(privileges, p)
+	}
+	return privileges, nil
+}
+
+// enableProcessPrivileges enables privileges globally for the process.
+func enableProcessPrivileges(names []string) error {
+	return enableDisableProcessPrivilege(names, SE_PRIVILEGE_ENABLED)
+}
+
+// DisableProcessPrivileges disables privileges globally for the process.
+func DisableProcessPrivileges(names []string) error {
+	return enableDisableProcessPrivilege(names, 0)
+}
+
+func enableDisableProcessPrivilege(names []string, action uint32) error {
+	privileges, err := mapPrivileges(names)
+	if err != nil {
+		return err
+	}
+
+	p := windows.CurrentProcess()
+	var token windows.Token
+	err = windows.OpenProcessToken(p, windows.TOKEN_ADJUST_PRIVILEGES|windows.TOKEN_QUERY, &token)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		_ = token.Close()
+	}()
+	return adjustPrivileges(token, privileges, action)
+}
+
+func adjustPrivileges(token windows.Token, privileges []uint64, action uint32) error {
+	var b bytes.Buffer
+	_ = binary.Write(&b, binary.LittleEndian, uint32(len(privileges)))
+	for _, p := range privileges {
+		_ = binary.Write(&b, binary.LittleEndian, p)
+		_ = binary.Write(&b, binary.LittleEndian, action)
+	}
+	prevState := make([]byte, b.Len())
+	reqSize := uint32(0)
+	success, err := adjustTokenPrivileges(token, false, &b.Bytes()[0], uint32(len(prevState)), &prevState[0], &reqSize)
+	if !success {
+		return err
+	}
+	if err == ERROR_NOT_ALL_ASSIGNED { //nolint:errorlint // err is Errno
+		return &PrivilegeError{privileges}
+	}
+	return nil
+}
+
+func getPrivilegeName(luid uint64) string {
+	var nameBuffer [256]uint16
+	bufSize := uint32(len(nameBuffer))
+	err := lookupPrivilegeName("", &luid, &nameBuffer[0], &bufSize)
+	if err != nil {
+		return fmt.Sprintf("<unknown privilege %d>", luid)
+	}
+
+	var displayNameBuffer [256]uint16
+	displayBufSize := uint32(len(displayNameBuffer))
+	var langID uint32
+	err = lookupPrivilegeDisplayName("", &nameBuffer[0], &displayNameBuffer[0], &displayBufSize, &langID)
+	if err != nil {
+		return fmt.Sprintf("<unknown privilege %s>", string(utf16.Decode(nameBuffer[:bufSize])))
+	}
+
+	return string(utf16.Decode(displayNameBuffer[:displayBufSize]))
+}
+
+func adjustTokenPrivileges(token windows.Token, releaseAll bool, input *byte, outputSize uint32, output *byte, requiredSize *uint32) (success bool, err error) {
```

Comment:
Probably, these were copied from winio.

Reply:
I've found the source file: https://github.com/microsoft/go-winio/blob/main/zsyscall_windows.go . _I don't want to see copied code ever again that does not clearly state where it originates from._ It's just ridiculous that I have to search for the original source of these functions.

Omitting that information also forces me to waste time on reviewing known good code.

To make matter worse, there have been a few small changes to those functions (apparently to fix linter warnings), which when compared to the latest upstream version of the autogenerated code seem to be buggy. See my other comments for details.

Reply:
That's a little harsh. I wrote this code 2 years ago before I decided to submit it back to restic.

The upstream code was changed only last month - https://github.com/microsoft/go-winio/commit/3c9576c9346a1892dee136329e7e15309e82fb4f
Eg.
	- r0, _, _ := syscall.Syscall(procHcsFormatWritableLayerVhd.Addr(), 1, uintptr(handle), 0, 0)
	+ r0, _, _ := syscall.SyscallN(procHcsFormatWritableLayerVhd.Addr(), uintptr(handle))
	
The changes I made 2 years ago were to fix the deprecations - 
Deprecated: Use SyscallN instead.	

Extra 0s at the end of SyscallN are ignored and hence this slip wasn't caught, and it also doesn't cause any bugs, though I agree it should be cleaned.

Also, I get your point on including links to the source and will try to do that, this was missed because the code was created from upstream a long time ago.

Reply:
> That's a little harsh.

Let's just say that I was rather annoyed after noticing that I have to diff the syscall wrappers and search for the corresponding source code.

> I wrote this code 2 years ago before I decided to submit it back to restic.

That perfectly explains why the code looks the way it does. Thanks for contributing those changes!